src/utils/sharedbuffergeometry.js

```diff
@@ -0,0 +1,135 @@
+export default class SharedBufferGeometry {
+  constructor(material, primitiveMode) {
+    this.material = material;
+    this.primitiveMode = primitiveMode;
+
+    this.maxBufferSize = 1000000;
+    this.geometries = [];
+    this.current = null;
+    this.drawing = new THREE.Object3D();
+    this.addBuffer(false);
+  }
+
+  restartPrimitive() {
+    if (this.idx.position >= this.current.attributes.position.count) {
+      this.addBuffer(false);
+    } else if (this.idx.position !== 0) {
+      let prev = (this.idx.position - 1) * 3;
+      const position = this.current.attributes.position.array;
+      this.addVertex(position[prev++], position[prev++], position[prev++]);
+
+      this.idx.color++;
+      this.idx.normal++;
+    }
+  }
+
+  remove(prevIdx, idx) {
+    // Loop through all the attributes: position, color, normal,...
+    if (this.idx.position > idx.position) {
+      for (const key in this.idx) {
+        const componentSize = 3;
+        let pos = prevIdx[key] * componentSize;
+        const start = (idx[key] + 1) * componentSize;
+        const end = this.idx[key] * componentSize;
+        for (let i = start; i < end; i++) {
+          this.current.attributes[key].array[pos++] = this.current.attributes[key].array[i];
+        }
+        const diff = idx[key] - prevIdx[key] + 1;
+        this.idx[key] -= diff;
+      }
+    } else {
+      for (const key in this.idx) {
+        const diff = idx[key] - prevIdx[key];
+        this.idx[key] -= diff;
+      }
+    }
+
+    this.update();
+  }
+
+  undo(prevIdx) {
+    this.idx = prevIdx;
+    this.update();
+  }
+
+  addBuffer(copyLast) {
+    const geometry = new THREE.BufferGeometry();
+
+    const vertices = new Float32Array(this.maxBufferSize * 3);
+    const normals = new Float32Array(this.maxBufferSize * 3);
+    const colors = new Float32Array(this.maxBufferSize * 3);
+
+    const mesh = new THREE.Mesh(geometry, this.material);
+
+    mesh.drawMode = this.primitiveMode;
+
+    mesh.frustumCulled = false;
+    mesh.vertices = vertices;
+
+    this.object3D = new THREE.Object3D();
```

Comment:
I think this can be `const object3D = ...` instead of `this.object3D`. I don't see this being referred to anywhere else.

Reply:
Ah yea, there's probably a few instances of stuff like this in this file since it came over from A-painter. I caught a few of them but thanks for finding this one.